timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -384,86 +420,116 @@ template void __global__ k_adjust_weights<double, 1>(
     }
 }
 
 template void __global__ k_set_sampled_weight_block<float, 512>(
     const int N,
+    const int batch_size,
     const int mol_size,
+    const int num_weights,
     const int *__restrict__ target_atoms,
     const float *__restrict__ per_atom_energies,
     const float inv_kT,
     __int128 *__restrict__ intermediate_accum);
 template void __global__ k_set_sampled_weight_block<double, 512>(
     const int N,
+    const int batch_size,
     const int mol_size,
+    const int num_weights,
     const int *__restrict__ target_atoms,
     const double *__restrict__ per_atom_energies,
     const double inv_kT,
     __int128 *__restrict__ intermediate_accum);
 
 template <typename RealType, int THREADS_PER_BLOCK>
 void __global__ k_set_sampled_weight_reduce(
+    const int batch_size,
+    const int num_weights,
     const int num_intermediates,
-    const int *__restrict__ samples,                 // [1]
-    const __int128 *__restrict__ intermediate_accum, // [num_intermediates]
-    RealType *__restrict__ log_weights) {
+    const int *__restrict__ samples,                 // [batch_size]
+    const __int128 *__restrict__ intermediate_accum, // [batch_size, num_intermediates]
+    RealType *__restrict__ log_weights               // [batch_size, num_weights]
+) {
     __shared__ __int128 accumulators[THREADS_PER_BLOCK];
 
-    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    // One block per set of weights
+    int idx_in_batch = blockIdx.y;
 
-    // Zero all of the accumulators
-    __int128 accumulator = 0;
-    while (idx < num_intermediates) {
-        accumulator += intermediate_accum[idx];
-        idx += gridDim.x * blockDim.x;
-    }
-    accumulators[threadIdx.x] = accumulator;
-    __syncthreads();
-    block_energy_reduce<THREADS_PER_BLOCK>(accumulators, threadIdx.x);
-    if (threadIdx.x == 0) {
-        int sample_idx = samples[0];
-        log_weights[sample_idx] =
-            fixed_point_overflow(accumulators[0]) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(accumulators[0]);
+    while (idx_in_batch < batch_size) {
+        int offset = idx_in_batch * num_intermediates;
+        // Threads act on a specific set of weights which is why just threadIdx.x
+        int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+        // Zero all of the accumulators
+        __int128 accumulator = 0;
+        while (idx < num_intermediates) {
+            accumulator += intermediate_accum[offset + idx];
+            idx += gridDim.x * blockDim.x;
+        }
+        accumulators[threadIdx.x] = accumulator;
+        __syncthreads();
+        block_energy_reduce<THREADS_PER_BLOCK>(accumulators, threadIdx.x);
+        if (threadIdx.x == 0) {
+            int mol_idx = samples[idx_in_batch];
+            log_weights[idx_in_batch * num_weights + mol_idx] =
+                fixed_point_overflow(accumulators[0]) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(accumulators[0]);
+        }
+
+        idx_in_batch += gridDim.y * blockDim.y;
```

Comment:
Not sure if intentional, but if this kernel is launched with:
```
    k_set_sampled_weight_reduce<RealType, WEIGHT_THREADS_PER_BLOCK>
        <<<dim3(1, batch_size_, 1), WEIGHT_THREADS_PER_BLOCK, 0, stream>>>(
```
and `static const int WEIGHT_THREADS_PER_BLOCK = 512; 
Then `gridDim.y == batchSize` but `blockDim.y == 1`?

Reply:
That is intentional. I had originally launched it with `<<<batch_size, WEIGHT_THREADS_PER_BLOCK, 0, stream>>>`, but seemed confusing as it made `blockIdx.x` mean something slightly differently.

Think that `gridDim.y == 1` and `blockDim.y == batchSize`, though would need to verify.

Reply:
Did you verify this in the end?

Reply:
I was incorrect, gridDim.y == batchSize and blockDim.y == 1. Sorry we had discussed this offline and forgot to comment here. The behavior is still as expected, just swapped the gridDim and blockDim in my mind.